### PR TITLE
[CI] Fix kernel execution in fuzztests and run fuzztests on L0 in CI

### PR DESCRIPTION
**Pull request by PatKamin**
Nightly run: https://github.com/oneapi-src/unified-runtime/actions/runs/9466770778

- Run fuzztests in CI on Level Zero instead of the Null Adapter
- Fixes kernel not triggering in fuzztests:
`Failed to open a kernel source file: /home/runner/work/unified-runtime/unified-runtime/build/test/conformance/device_binaries/bar/sycl_spir641.spv`
- Fixes short fuzztests failing to create a device on Level Zero
- Fuzztests refactoring
- Enable to input any custom paths to external level zero
- Build level zero with gcc for fuzztests to avoid crashes when binary finishes
- Verbose long fuzztests (easier to spot if fuzztests actually get beyond urPlatformGet)

### CODE REVIEW

File: .github/workflows/build-fuzz-reusable.yml [hunk 1]
```diff
@@ -0,0 +1,74 @@
+---
+name: Build - Fuzztests on L0 HW - Reusable
+
+on:
+  workflow_call:
+    inputs:
+      test_name:
```
Comment (lukaszstolarczuk) [inline .github/workflows/build-fuzz-reusable.yml:7]: `test_label` to be precise

File: test/fuzz/CMakeLists.txt [hunk 2]
```diff
@@ -33,17 +22,42 @@ function(add_fuzz_test name label)
...
+            list(APPEND ENV_VARS
+                    NEOReadDebugKeys=1
+                    DisableDeepBind=1)
+        endif()
+    else()
+        list(APPEND ENV_VARS UR_ADAPTERS_FORCE_LOAD=\"$<TARGET_FILE:ur_adapter_null>\")
+    endif()
+    
+    add_test(NAME ${TEST_TARGET_NAME}
+        COMMAND fuzztest-base ${ARGN} -verbosity=1 -detect_leaks=0
```
Comment (lukaszstolarczuk) [inline test/fuzz/CMakeLists.txt:39]: is this `-detect_leaks=0` needed here? it's set for ASAN separately

File: test/fuzz/CMakeLists.txt [hunk 3]
```diff
@@ -5,26 +5,15 @@
```
Comment (lukaszstolarczuk) [inline test/fuzz/CMakeLists.txt:5]: pls bump copyright dates (in all files)

**Review by lukaszstolarczuk — COMMENTED** [review]
pls rebase

File: test/fuzz/CMakeLists.txt (hunk 2 above)
Comment (PatKamin) [reply test/fuzz/CMakeLists.txt:39]: It's required only here. I've removed the env var now.

File: .github/workflows/build-fuzz-reusable.yml (hunk 1 above)
Comment (PatKamin) [reply .github/workflows/build-fuzz-reusable.yml:7]: Yes, that fits better, thanks.

File: test/fuzz/CMakeLists.txt (hunk 3 above)
Comment (PatKamin) [reply test/fuzz/CMakeLists.txt:5]: Thanks, done

File: .github/workflows/build-fuzz-reusable.yml (hunk 1 above)
Comment (lukaszstolarczuk) [reply .github/workflows/build-fuzz-reusable.yml]: heh, pls fix also usage of this var, now 😉

Comment (PatKamin) [reply .github/workflows/build-fuzz-reusable.yml]: Done

**Review by kbenzie — APPROVED** [review]

File: .github/workflows/build-fuzz-reusable.yml [hunk 4]
```diff
@@ -0,0 +1,74 @@
...
+jobs:
+  fuzztest-build-hw:
+    name: Build and run fuzz tests on L0 HW
+    if: github.repository == 'oneapi-src/unified-runtime'  # run only on upstream; forks won't have the HW
+    strategy:
+      matrix:
+        build_type: [Debug, Release]
+        compiler: [{c: clang, cxx: clang++}]
+
+    runs-on: 'FUZZTESTS'
```
Comment (pbalcer) [inline .github/workflows/build-fuzz-reusable.yml:23]: add `# TODO: fix runners / OS / whatever`.

File: .github/workflows/build-fuzz-reusable.yml [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+
+    - name: Build level zero with gcc
+      run: |
+        git clone -b v1.16.1 --depth=1 https://github.com/oneapi-src/level-zero.git ${{github.workspace}}/level-zero
+        cd ${{github.workspace}}/level-zero
+        cmake -B build -DCMAKE_BUILD_TYPE=${{matrix.build_type}} -DCMAKE_C_COMPILER=gcc -DCMAKE_CXX_COMPILER=g++
+        cmake --build build -j $(nproc)
+
+    - name: Configure CMake
+      # In order to use sanitizers, vm.mmap_rnd_bits=28 must be set in the system,
```
Comment (pbalcer) [inline .github/workflows/build-fuzz-reusable.yml:45]: this is not related to cmakes configure step. Move this to next to runs-on.

File: test/fuzz/urFuzz.cpp [hunk 6]
```diff
@@ -353,36 +353,53 @@ int ur_program_create_with_il(TestState &state) {
...
-    ur_event_handle_t event = nullptr;
+    ur_program_handle_t program;
+    ur_kernel_handle_t kernel;
+    ur_queue_handle_t queue;
+    ur_event_handle_t event;
     auto &context = state.contexts[state.context_num]->handle;
     auto &device = state.devices[state.device_num];
     std::string kernel_name =
-        uur::device_binaries::program_kernel_map["bar"][0];
+        uur::device_binaries::program_kernel_map["fill"][0];
```
Comment (pbalcer) [inline test/fuzz/urFuzz.cpp:363]: we need some generic utility to retrieve/use kernels in tests. Hardcoding this is not ideal, but will do for now...

**Review by pbalcer — APPROVED** [review]
just a few nits.

File: .github/workflows/build-fuzz-reusable.yml (hunk 5 above)
Comment (PatKamin) [reply .github/workflows/build-fuzz-reusable.yml:45]: Done

File: test/fuzz/urFuzz.cpp (hunk 6 above)
Comment (PatKamin) [reply test/fuzz/urFuzz.cpp:363]: Added TODO to remember this

**Comment by PatKamin** [comment]
Bumped level-zero used by tests to v1.17.6

**Comment by kbenzie** [comment]
> Bumped level-zero used by tests to v1.17.6

I wondered if it was something like that, thanks.
